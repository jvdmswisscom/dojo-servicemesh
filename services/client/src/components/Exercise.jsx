import React from 'react';
import PropTypes from 'prop-types';
import AceEditor from 'react-ace';
import 'brace/mode/python';
import 'brace/theme/solarized_dark';

const Exercise = (props) => {
  return (
    <div key={props.exercise.id}>
      <h5 className="title is-5">{props.exercise.body}</h5>
      <AceEditor
        mode="python"
        theme="solarized_dark"
        name={(props.exercise.id).toString()}
        fontSize={14}
        height={'175px'}
        showPrintMargin={true}
        showGutter={true}
        highlightActiveLine={true}
        value={props.editor.value}
        style={{
          marginBottom: '10px'
        }}
        editorProps={{
          $blockScrolling: Infinity
        }}
        onChange={props.onChange}
      />
      {props.isAuthenticated &&
        <div>
          <button
            className="button is-primary"
            onClick={(evt) => props.submitExercise(evt, props.exercise.id)}
            disabled={props.editor.button.isDisabled}
          >Run Code</button>
          {props.editor.showGrading &&
            <h5 className="title is-5">
              <span className="icon is-large">
                <i className="fas fa-spinner fa-pulse"></i>
              </span>
              <span className="grade-text">Grading...</span>
            </h5>
          }
          {props.editor.showCorrect &&
            <h5 className="title is-5">
              <span className="icon is-large">
                <i className="fas fa-check"></i>
              </span>
              <span className="grade-text">Correct!</span>
            </h5>
          }
          {props.editor.showIncorrect &&
            <h5 className="title is-5">
              <span className="icon is-large">
                <i className="fas fa-times"></i>
              </span>
              <span className="grade-text">Incorrect!</span>
            </h5>
          }
        </div>
      }
      <br/><br/>
    </div>
  )
};

Exercise.propTypes = {
  exercise: PropTypes.shape({
    body: PropTypes.string.isRequired,
    id: PropTypes.number.isRequired,
    test_code: PropTypes.string.isRequired,
    test_code_solution: PropTypes.string.isRequired,
  }).isRequired,
  editor: PropTypes.shape({
    button: PropTypes.object.isRequired,
    showCorrect: PropTypes.bool.isRequired,
    showGrading: PropTypes.bool.isRequired,
    showIncorrect: PropTypes.bool.isRequired,
    value: PropTypes.string.isRequired,
  }).isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
  submitExercise: PropTypes.func.isRequired,
};

export default Exercise;
