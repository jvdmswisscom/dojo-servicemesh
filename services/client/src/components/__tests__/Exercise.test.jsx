import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import AceEditor from 'react-ace';
jest.mock('react-ace');

import Exercise from '../Exercise';

const testData = {
  exercise: {
    id: 0,
    body: 'Define a function called sum that takes two integers as arguments and returns their sum.',
    test_code: 'sum(2,2)',
    test_code_solution: '4'
  },
  editor: {
    value: '# Enter your code here.',
    button: {
      isDisabled: false,
    },
    showGrading: false,
    showCorrect: false,
    showIncorrect: false,
  },
  isAuthenticated: false,
  onChange: jest.fn(),
  submitExercise: jest.fn(),
}

beforeEach(() => {
  console.error = jest.fn();
  console.error.mockClear();
});

test('Exercise renders properly', () => {
  const wrapper = shallow(<Exercise {...testData}/>);
  const heading = wrapper.find('h5');
  expect(heading.length).toBe(1);
  expect(heading.text()).toBe(testData.exercise.body)
  expect(console.error).toHaveBeenCalledTimes(0);
});

test('Exercise does not render properly when not all props are defined', () => {
  delete testData.submitExercise
  const wrapper = shallow(<Exercise {...testData}/>);
  const heading = wrapper.find('h5');
  expect(heading.length).toBe(1);
  expect(heading.text()).toBe(testData.exercise.body)
  expect(console.error).toHaveBeenCalledTimes(1);
});

test('Exercises renders a snapshot properly when not authenticated', () => {
  const tree = renderer.create(<Exercise {...testData}/>).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Exercises renders a snapshot properly when authenticated', () => {
  testData.isAuthenticated = true;
  const tree = renderer.create(<Exercise {...testData}/>).toJSON();
  expect(tree).toMatchSnapshot();
});
