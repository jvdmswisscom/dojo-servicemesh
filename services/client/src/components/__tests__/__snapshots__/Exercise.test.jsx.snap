// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exercises renders a snapshot properly when authenticated 1`] = `
<div>
  <h5
    className="title is-5"
  >
    Define a function called sum that takes two integers as arguments and returns their sum.
  </h5>
  <div>
    <button
      className="button is-primary"
      disabled={false}
      onClick={[Function]}
    >
      Run Code
    </button>
  </div>
  <br />
  <br />
</div>
`;

exports[`Exercises renders a snapshot properly when not authenticated 1`] = `
<div>
  <h5
    className="title is-5"
  >
    Define a function called sum that takes two integers as arguments and returns their sum.
  </h5>
  <br />
  <br />
</div>
`;
